### PR TITLE
perf: make tests run faster

### PR DESCRIPTION
**Pull request by petar-dambovaliev**
solves [issue](https://github.com/gnolang/gno/issues/1406)

This PR enables running tests in parallel by making the `Machine` data structure copy-able and thus threadsafe.

`DeepCopy` methods added to all types that are contained within the `Machine`. including types that implement interfaces `Expr`, `Object`, `Store` and  `Type`.

* Before each test is launched, the `Machine` object is deep copied and passed along so that test can read and write on its own instance.

* Global debugging functionality, including the accumulation of errors was moved into local scopes.

* the `var preprocessing int` counter was moved into its own type with the relevant methods and its passed as an argument

* Debugging is a field of every struct that needs to use it (i tried adding it as a method argument first, it didn't end well. The code was becoming extremely noisy)

### CODE REVIEW

**Comment by thehowl** [comment]
@petar-dambovaliev ready for review?

I'll try to take a look at it this week. (It's close to the top of my review-list, so should be able to take a look at it, but no guarantees).

Thanks for the hard work.

**Comment by zivkovicmilos** [comment]
> @petar-dambovaliev ready for review?
> 
> I'll try to take a look at it this week. (It's close to the top of my review-list, so should be able to take a look at it, but no guarantees).
> 
> Thanks for the hard work.

Should be ready for viewing as far as I know

File: gnovm/pkg/gnolang/debug.go [hunk 1]
```diff
@@ -18,15 +17,13 @@ import (
 // so it is still faster to first check the truth value
 // before calling debug.Println or debug.Printf.
 
-type debugging bool
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/debug.go:21]: jic9nec7mae31

cc @ajnavarro

**Review by zivkovicmilos — APPROVED** [review]
Ran this locally, seems to shave ~1min on my machine for `make test` in the `gnovm` subfolder 🎉 

Thank you for removing globals. I see this PR as a good start 👍

**Review by ajnavarro — COMMENTED** [review]
doing a partial review, GitHub is failing to load the diff.

File: gnovm/pkg/gnolang/values.go
Comment (thehowl) [inline gnovm/pkg/gnolang/values.go:1]: So, I don't think we necessarily need what @ajnavarro mentioned (ie. making `debugging` a singleton like it was before the PR), but I do think we need to split its two uses, which actually completely avoids what happened in this file: that every file type now gets the debugging variable.

I suggest that for all cases where the debugging variable guards an additional check like the vast majority of the ones here, that we instead opt for the same solution I'm proposing to disable debugging at compile time: build tags. These are all checks which provide "additional safety", so I would have a value like `const Safety = true`, enabled by default, that can be disabled with build tag `-tags gno_optimize` or `-tags gno_nosafety`.

**Review by thehowl — COMMENTED** [review]
So, I had a feeling that adding new pointers to all of these structs may actually not be the best thing for performance. The tests may run faster, but overall execution is actually _slower_:

```
goos: linux
goarch: amd64
pkg: github.com/gnolang/gno/gnovm/pkg/gnolang
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                                │ bench_master.txt │           bench_1406.txt            │
                                │      sec/op      │   sec/op     vs base                │
Benchdata/fib.gno_param:4-16          32.27µ ±  5%   32.47µ ± 7%        ~ (p=0.393 n=10)
Benchdata/fib.gno_param:8-16          228.8µ ±  5%   236.0µ ± 1%   +3.15% (p=0.000 n=10)
Benchdata/fib.gno_param:16-16         11.02m ±  4%   10.91m ± 5%        ~ (p=0.481 n=10)
Benchdata/loop.gno-16                 493.3n ± 11%   698.5n ± 6%  +41.60% (p=0.000 n=10)
Benchdata/matrix.gno_param:3-16       653.9µ ±  4%   707.7µ ± 6%   +8.22% (p=0.004 n=10)
Benchdata/matrix.gno_param:4-16       1.862m ±  3%   2.081m ± 4%  +11.74% (p=0.000 n=10)
Benchdata/matrix.gno_param:5-16       6.822m ±  6%   7.643m ± 5%  +12.04% (p=0.000 n=10)
Benchdata/matrix.gno_param:6-16       34.66m ±  5%   38.16m ± 5%  +10.11% (p=0.000 n=10)
geomean                               572.6µ         630.8µ       +10.17%
```

```
go test -v -bench 'Benchdata' -run 'NONE' -count 10 | tee ~/throw/bench_1406.txt
```

These are tests done with the benchmarking system I added in #1624 (note it specifically targets benchmarking execution, not preprocessing) and as you can see there is an average 10% more time taken on each operation.

I think this makes clear our need for good benchmarking and proves the point of setting this up in #1624. Please, if you see problems with the methodology, or see a benchmark which disproves these benchmarks, do point them out. I pushed the branch with this branch + the changes in #1624 in [`1406-with-bench`](https://github.com/gnolang/gno/tree/1406-with-bench)

I think there are some foundational ideas in this PR which are good, some which need to be iterated before we can have tests which are both parallelized but don't sacrifice execution-time performance.

Thanks!

**Review by thehowl — CHANGES_REQUESTED** [review]
Sorry, meant to indicate request changes.

File: gnovm/pkg/gnolang/nodes_string.go [hunk 2]
```diff
@@ -123,7 +123,7 @@ func (x IndexExpr) String() string {
 }
 
 func (x SelectorExpr) String() string {
-	// NOTE: for debugging selector issues:
+	// NOTE: for  selector issues:
```
[the remark this answers is not in the document]
Comment (petar-dambovaliev) [reply gnovm/pkg/gnolang/nodes_string.go:126]: intellij find and replace

File: gnovm/Makefile [hunk 3]
```diff
@@ -39,7 +39,7 @@ fmt:
 .PHONY: test
 test: _test.cmd _test.pkg _test.gnolang
 
-GOTEST_FLAGS ?= -v -p 1 -timeout=30m
+GOTEST_FLAGS ?= -timeout=30m
```
[the remark this answers is not in the document]
Comment (petar-dambovaliev) [reply gnovm/Makefile:42]: It is giving out a lot of spam. You generally don't care for verbose when running all tests.
You care about verbose when running a single test.

**Comment by petar-dambovaliev** [comment]
https://github.com/gnolang/gno/pull/1417#discussion_r1476316421

2 threads appending a slice without a synchronization mechanism is a concurrency violation.

**Pull request closed without merging**